pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,386 @@
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning"
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning/scheduling"
+	"github.com/aws/karpenter-core/pkg/operator/controller"
+
+	"github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	disruptionevents "github.com/aws/karpenter-core/pkg/controllers/disruption/events"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/utils/nodeclaim"
+)
+
+const (
+	queueBaseDelay   = 1 * time.Second
+	queueMaxDelay    = 10 * time.Second
+	maxRetryDuration = 10 * time.Minute
+)
+
+type Command struct {
+	ReplacementKeys   []*NodeClaimKey
+	Candidates        []*state.StateNode
+	ConsolidationType string    // used for metrics
+	Method            string    // used for metrics
+	Reason            string    // used for metrics
+	TimeAdded         time.Time // timeAdded is used to track timeouts
+	LastError         error
+}
+
+// NodeClaimKey wraps a nodeclaim.Key with an initialized field to save on readiness checks and identify
+// when a nodeclaim is first initialized for metrics and events.
+type NodeClaimKey struct {
+	nodeclaim.Key
+	// Use a bool track if a node has already been initialized so we can fire metrics for intialization once.
+	// This intentionally does not capture nodes that go initialized then go NotReady after as other pods can
+	// schedule to this node as well.
+	Initialized bool
+}
+
+type UnrecoverableError struct {
+	error
+}
+
+func NewUnrecoverableError(err error) *UnrecoverableError {
+	return &UnrecoverableError{error: err}
+}
+
+func IsUnrecoverableError(err error) bool {
+	if err == nil {
+		return false
+	}
+	var unrecoverableError *UnrecoverableError
+	return errors.As(err, &unrecoverableError)
+}
+
+type Queue struct {
+	workqueue.RateLimitingInterface
+	// providerID -> command, maps a candidate to its command. Each command has a list of candidates that can be used
+	// to map to itself.
+	ProviderIDToCommand map[string]*Command
+
+	kubeClient  client.Client
+	recorder    events.Recorder
+	cluster     *state.Cluster
+	clock       clock.Clock
+	provisioner *provisioning.Provisioner
+}
+
+// NewQueue creates a queue that will asynchronously orchestrate disruption commands
+func NewQueue(kubeClient client.Client, recorder events.Recorder, cluster *state.Cluster, clock clock.Clock,
+	provisioner *provisioning.Provisioner) *Queue {
+	queue := &Queue{
+		RateLimitingInterface: workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(queueBaseDelay, queueMaxDelay)),
```

Comment:
It's going to regress to 10s when it's retrying the same command, no? I think that's the expected behavior. I could imagine no replacement needed, but maybe there's a fleeting kubeClient error where i'd want the backoff to be smaller

Reply:
Well, the back-off on retry means that it will keep going to 10s even if there is no transient error, right? Or am I misunderstanding the back-off logic?